pkg/state/state_nic_feature_discovery_test.go

```diff
@@ -0,0 +1,180 @@
+	))
+
+	// Container Resources
+	Expect(template.Spec.Containers[0].Resources.Limits).To(Equal(image.ContainerResources[0].Limits))
+	Expect(template.Spec.Containers[0].Resources.Requests).To(Equal(image.ContainerResources[0].Requests))
+}
+
+func assertCommonDeploymentFields(u *unstructured.Unstructured, image *mellanoxv1alpha1.ImageSpec) {
+	d := &appsv1.Deployment{}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.UnstructuredContent(), d)
+	Expect(err).ToNot(HaveOccurred())
+	assertCommonPodTemplateFields(&d.Spec.Template, image)
+}
+
+func assertCommonDaemonSetFields(u *unstructured.Unstructured,
+	image *mellanoxv1alpha1.ImageSpec, policy *mellanoxv1alpha1.NicClusterPolicy) {
+	ds := &appsv1.DaemonSet{}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.UnstructuredContent(), ds)
+	Expect(err).ToNot(HaveOccurred())
+	assertCommonPodTemplateFields(&ds.Spec.Template, image)
+	// Tolerations
+	Expect(ds.Spec.Template.Spec.Tolerations).To(ContainElements(
+		corev1.Toleration{Key: "first-taint"},
+		corev1.Toleration{
+			Key:               "nvidia.com/gpu",
+			Operator:          "Exists",
+			Value:             "",
+			Effect:            "NoSchedule",
+			TolerationSeconds: nil,
+		},
+	))
+
+	// NodeAffinity
+	Expect(ds.Spec.Template.Spec.Affinity.NodeAffinity).To(Equal(policy.Spec.NodeAffinity))
+}
+
+func getTestImageSpec() *mellanoxv1alpha1.ImageSpec {
+	return &mellanoxv1alpha1.ImageSpec{
+		Image:            "image-one",
+		Repository:       "repository",
+		Version:          "five",
+		ImagePullSecrets: []string{"secret-one", "secret-two"},
+	}
+}
+
+func addContainerResources(imageSpec *mellanoxv1alpha1.ImageSpec,
+	containerName, requestValue, limitValue string) *mellanoxv1alpha1.ImageSpec {
+	i := imageSpec.DeepCopy()
+	i.ContainerResources = append(i.ContainerResources, []mellanoxv1alpha1.ResourceRequirements{
+		{
+			Name:     containerName,
+			Limits:   map[corev1.ResourceName]resource.Quantity{"resource-one": resource.MustParse(limitValue)},
+			Requests: map[corev1.ResourceName]resource.Quantity{"resource-one": resource.MustParse(requestValue)},
+		},
+	}...)
+	return i
+}
+
+func isNamespaced(obj *unstructured.Unstructured) bool {
+	return obj.GetKind() != "CustomResourceDefinition" &&
+		obj.GetKind() != "ClusterRole" &&
+		obj.GetKind() != "ClusterRoleBinding" &&
+		obj.GetKind() != "ValidatingWebhookConfiguration"
+}
+
+func assertCNIBinDirForDS(u *unstructured.Unstructured) {
+	ds := &appsv1.DaemonSet{}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.UnstructuredContent(), ds)
+	Expect(err).ToNot(HaveOccurred())
+	for i := range ds.Spec.Template.Spec.Volumes {
+		vol := ds.Spec.Template.Spec.Volumes[i]
+		if vol.Name == "cnibin" {
+			Expect(vol.HostPath).NotTo(BeNil())
+			Expect(vol.HostPath.Path).To(Equal("custom-cni-bin-directory"))
+		}
+	}
+}
+
+func GetManifestObjectsTest(ctx context.Context, cr *mellanoxv1alpha1.NicClusterPolicy, catalog state.InfoCatalog,
```

Comment:
I think we could expand this test to include more objects and checks. We'd then be able to run it on all Manifests for templating which is in common. Specific tests can then be run for fields which are different or novel between states.

Reply:
I like the idea. This function is already used in two files. Should we put it to common_test.go?

Reply:
+1, this can be done in a followup.